src/Analyzers/CSharp/CodeFixes/UseCollectionInitializer/CSharpUseCollectionInitializerCodeFixProvider.cs

```diff
@@ -36,80 +42,187 @@ public CSharpUseCollectionInitializerCodeFixProvider()
+            foreach (var match in matches)
+            {
+                var expression = GetExpression(match);
+                totalLength += expression.Span.Length;
+                totalLength += ", ".Length;
 
-                var newExpression = ConvertExpression(expressionStatement.Expression)
-                    .WithoutTrivia()
-                    .WithPrependedLeadingTrivia(newTrivia);
+                if (totalLength > wrappingLength)
+                    return true;
+            }
 
-                if (i < matches.Length - 1)
+            return false;
+
+            static ExpressionSyntax GetExpression(Match<StatementSyntax> match)
+                => match.Statement switch
                 {
-                    nodesAndTokens.Add(newExpression);
-                    var commaToken = SyntaxFactory.Token(SyntaxKind.CommaToken)
-                        .WithTriviaFrom(expressionStatement.SemicolonToken);
+                    ExpressionStatementSyntax expressionStatement => expressionStatement.Expression,
+                    ForEachStatementSyntax foreachStatement => foreachStatement.Expression,
+                    _ => throw ExceptionUtilities.Unreachable(),
+                };
+        }
+
+        public static SeparatedSyntaxList<TNode> AddLineBreaks<TNode>(
+            SeparatedSyntaxList<TNode> nodes, bool includeFinalLineBreak)
+            where TNode : SyntaxNode
+        {
+            using var _ = ArrayBuilder<SyntaxNodeOrToken>.GetInstance(out var nodesAndTokens);
 
-                    nodesAndTokens.Add(commaToken);
+            var nodeOrTokenList = nodes.GetWithSeparators();
+            foreach (var item in nodeOrTokenList)
+            {
+                var addLineBreak = item.IsToken || (includeFinalLineBreak && item == nodeOrTokenList.Last());
+                if (addLineBreak && item.GetTrailingTrivia() is not [.., (kind: SyntaxKind.EndOfLineTrivia)])
+                {
+                    nodesAndTokens.Add(item.WithAppendedTrailingTrivia(ElasticCarriageReturnLineFeed));
                 }
                 else
                 {
-                    newExpression = newExpression.WithTrailingTrivia(
-                        expressionStatement.GetTrailingTrivia());
-                    nodesAndTokens.Add(newExpression);
+                    nodesAndTokens.Add(item);
                 }
             }
 
-            return SyntaxFactory.SeparatedList<ExpressionSyntax>(nodesAndTokens);
+            return SeparatedList<TNode>(nodesAndTokens);
         }
 
-        private static ExpressionSyntax ConvertExpression(ExpressionSyntax expression)
+        private static SeparatedSyntaxList<TElement> CreateElements<TElement>(
+            BaseObjectCreationExpressionSyntax objectCreation,
+            ImmutableArray<Match<StatementSyntax>> matches,
+            Func<Match<StatementSyntax>?, ExpressionSyntax, TElement> createElement)
+            where TElement : SyntaxNode
```

Comment:
this is similar cod eto what we had before.  Except:

1. newline handling is pulled out of this to the callers.
2. this is generified so it can make expression lists (for collection initializers) or element lists (for collection expressions).